builder.go

```diff
@@ -0,0 +1,235 @@
+package biscuit
+
+import (
+	"errors"
+	"io"
+
+	"github.com/flynn/biscuit-go/datalog"
+	"github.com/flynn/biscuit-go/pb"
+	"github.com/flynn/biscuit-go/sig"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	ErrDuplicateFact     = errors.New("biscuit: fact already exists")
+	ErrInvalidBlockIndex = errors.New("biscuit: invalid block index")
+)
+
+type Builder interface {
+	AddAuthorityFact(fact *Fact) error
+	AddAuthorityRule(rule *Rule) error
+	AddAuthorityCaveat(rule *Rule) error
+	AddRight(resource, right string) error
+	Build() (*Biscuit, error)
+}
+
+type builder struct {
+	rng  io.Reader
+	root sig.Keypair
+
+	symbolsStart int
+	symbols      *datalog.SymbolTable
+	facts        *datalog.FactSet
+	rules        []*datalog.Rule
+	caveats      []*datalog.Caveat
+	context      string
+}
+
+func NewBuilder(rng io.Reader, root sig.Keypair) Builder {
+	return BuilderWithSymbols(rng, root, DefaultSymbolTable)
+}
+
+func BuilderWithSymbols(rng io.Reader, root sig.Keypair, symbols *datalog.SymbolTable) Builder {
+	return &builder{
+		rng:  rng,
+		root: root,
+
+		symbolsStart: symbols.Len(),
+		symbols:      symbols,
+		facts:        new(datalog.FactSet),
+	}
+}
+
+func (b *builder) AddAuthorityFact(fact *Fact) error {
+	if len(fact.Predicate.IDs) == 0 || fact.Predicate.IDs[0] != SymAuthority {
+		fact.Predicate.IDs[0] = SymAuthority
+	}
+
+	dlFact := fact.convert(b.symbols)
+	if !b.facts.Insert(dlFact) {
+		return ErrDuplicateFact
+	}
+
+	return nil
+}
+
+func (b *builder) AddAuthorityRule(rule *Rule) error {
+	if len(rule.Head.IDs) == 0 || rule.Head.IDs[0] != SymAuthority {
+		rule.Head.IDs[0] = SymAuthority
+	}
+
+	dlRule := rule.convert(b.symbols)
+	b.rules = append(b.rules, &dlRule)
+	return nil
+}
+
+func (b *builder) AddAuthorityCaveat(rule *Rule) error {
+	b.caveats = append(b.caveats, &datalog.Caveat{Queries: []datalog.Rule{rule.convert(b.symbols)}})
+	return nil
+}
+
+func (b *builder) AddRight(resource, right string) error {
+	return b.AddAuthorityFact(&Fact{
+		Predicate: Predicate{
+			Name: "right",
+			IDs: []Atom{
+				SymAuthority,
+				String(resource),
+				Symbol(right),
+			},
+		},
+	})
+}
+
+func (b *builder) Build() (*Biscuit, error) {
+	return New(b.rng, b.root, b.symbols, &Block{
+		index:   0,
+		symbols: b.symbols.SplitOff(b.symbolsStart),
+		facts:   b.facts,
+		rules:   b.rules,
+		caveats: b.caveats,
+		context: b.context,
+	})
+}
+
+func From(serialized []byte) (*Biscuit, error) {
+	return FromWithSymbols(serialized, DefaultSymbolTable)
+}
+
+func FromWithSymbols(serialized []byte, symbols *datalog.SymbolTable) (*Biscuit, error) {
```

Comment:
Do you think it makes more sense to have an `Unmarshaler` struct that has a `Symbols` field? That's a pretty standard Go convention that allows adding more options later.

Reply:
What do you think making it a `Marshaller`  and move the `Biscuit.Serialize`  on it too ?

https://github.com/flynn/biscuit-go/blob/1486c9c8266ff40dbb2a6f5cf48bedc569004af9/biscuit.go#L160-L162 

I like getting all the marshalling stuff on the same struct, but unsure with the `biscuit.container` private field in the way.